### ISSUE TITLE
[bug] TCH004 false positive with function-scoped imports

### ISSUE THREAD

**Issue by smackesey** [charliermarsh/ruff#4941]
I was unable to create a simple reproduction of this, but TCH004 on ruff 0.0.271 gives several false positives in Dagster when there are function-local imports. Running ruff on https://github.com/dagster-io/dagster/blame/master/python_modules/dagster/dagster/_core/definitions/op_definition.py will give this error:

```
python_modules/dagster/dagster/_core/definitions/op_definition.py:53:42: TCH004 [*] Move import `.decorators.op_decorator.DecoratedOpFunction` out of type-checking block. Import is used for more than type hinting.
```

But this isn't the case-- all the places where `DecoratedOpFunction` is used outside of type annotations, there is a function-local import covering it. When ruff autofixes this it pulls the import out of the `TYPE_CHECKING` block which then causes a circular import error.

**Comment by charliermarsh** [comment]
Thanks, that's a helpful example.

**Comment by charliermarsh** [comment]
Oh, hah, this was also fixed by #4942.
